packages/core/mesh/edge-client/src/persistent-lifecycle.ts

```diff
@@ -0,0 +1,105 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { DeferredTask, sleep } from '@dxos/async';
+import { cancelWithContext, LifecycleState, Resource } from '@dxos/context';
+import { warnAfterTimeout } from '@dxos/debug';
+import { log } from '@dxos/log';
+
+const DEFAULT_RESTART_DELAY = 100;
+const DEFAULT_MAX_RESTART_DELAY = 5000;
+
+export type PersistentLifecycleParams = {
+  /**
+   * Create connection.
+   * If promise resolves successfully, connection is considered established.
+   */
+  start: () => Promise<void>;
+
+  /**
+   * Reset connection to initial state.
+   */
+  stop: () => Promise<void>;
+
+  /**
+   * Called after successful start.
+   */
+  onRestart?: () => Promise<void>;
+
+  /**
+   * Maximum delay between restartion attempts.
+   * Default: 5000ms
+   */
+  maxRestartDelay?: number;
+};
+
+/**
+ * Handles restarts (e.g. persists connection).
+ * Restarts are scheduled with exponential backoff.
+ */
+export class PersistentLifecycle extends Resource {
+  private readonly _start: () => Promise<void>;
+  private readonly _stop: () => Promise<void>;
+  private readonly _onRestart?: () => Promise<void>;
+  private readonly _maxRestartDelay: number;
+
+  private _restartTask?: DeferredTask = undefined;
+  private _restartAfter = 0;
+
+  constructor({ start, stop, onRestart, maxRestartDelay = DEFAULT_MAX_RESTART_DELAY }: PersistentLifecycleParams) {
```

Comment:
consider having this manage initial start/stop too. It seems right now this is used for reastarts, but edge-client also creates and destroys the websocket manually too